src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,688 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javax.crypto;
+
+import jdk.internal.javac.PreviewFeature;
+import sun.security.jca.GetInstance;
+import sun.security.jca.GetInstance.Instance;
+import sun.security.util.Debug;
+
+import java.security.InvalidAlgorithmParameterException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.Security;
+import java.security.Provider;
+import java.security.Provider.Service;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.Iterator;
+import java.util.Objects;
+
+/**
+ * This class provides the functionality of a Key Derivation Function (KDF),
+ * which is a cryptographic algorithm for deriving additional keys from input
+ * keying material and (optionally) other data.
+ * <p>
+ * {@code KDF} objects are instantiated with the {@code getInstance} family
+ * of methods.
+ * <p>
+ * The class has two derive methods, {@code deriveKey} and {@code deriveData}.
+ * The {@code deriveKey} method accepts an algorithm name and returns a
+ * {@code SecretKey} object with the specified algorithm. The {@code deriveData}
+ * method returns a byte array of raw data.
+ * <p>
+ * API Usage Example:
+ * {@snippet lang = java:
+ *    KDF kdfHkdf = KDF.getInstance("HKDF-SHA256");
+ *
+ *    AlgorithmParameterSpec derivationSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *    SecretKey sKey = kdfHkdf.deriveKey("AES", derivationSpec);
+ *}
+ * <br>
+ * <h2><a id="ConcurrentAccess">Concurrent Access</a></h2>
+ * Unless otherwise documented by an implementation, the methods defined in
+ * this class are not thread-safe. Multiple threads that need to access a
+ * single object concurrently should synchronize amongst themselves and
+ * provide the necessary locking. Multiple threads each manipulating separate
+ * objects need not synchronize.
+ * <br>
+ * <h2><a id="DelayedProviderSelection">Delayed Provider Selection</a></h2>
+ * If a provider is not specified when calling one of the {@code getInstance}
+ * methods, the implementation delays the selection of the provider until the
+ * {@code deriveKey} or {@code deriveData} method is called. This is called
+ * <i>delayed provider selection</i>. The primary reason this is done is to
+ * ensure that the selected provider can handle the key material that is passed
+ * to those methods - for example, the key material may reside on a
+ * hardware device that only a specific {@code KDF} provider can utilize.
```

Comment:
I think we should also add this sentence. This is similar to the provider search algorithm that is documented in all current JCE/JCA getInstance APIs so it is important to repeat it here.

"Once initiated, the selection process traverses the list of registered security providers, starting with the most preferred `Provider`. A new KDF object encapsulating the `KDFSpi` implementation from the first provider that supports the specified algorithm and optional parameters is returned."